### ISSUE TITLE
DWARFExpression::DumpLocation doesn't respect register kind when dumping.

### ISSUE THREAD

**Issue by ZequanWu** [llvm/llvm-project#53575]
For example, if the register kind is `eRegisterKindLLDB`, it will dump wrong register enums. [`DWARFExpression::DumpLocation`](https://github.com/llvm/llvm-project/blob/main/lldb/source/Expression/DWARFExpression.cpp#L90) should convert the data from `m_reg_kind` to dwarf kind before dumping if it's not already. But that seems to require a register context to do so.

**Comment by labath** [comment]
This is probably a side-effect of switching to llvm's DWARFExpression class for dumping.

Another way of looking at this/fixing it would be to construct the expression using the dwarf register numbers in the first place. I can't think of any strong reason for why we would need the ability to construct these expressions using other register enums.

**Comment by ZequanWu** [comment]
If we want to use dwarf register numbers in first place, then we may need something similar to https://github.com/llvm/llvm-project/blob/main/lldb/source/Plugins/SymbolFile/NativePDB/CodeViewRegisterMapping.cpp#L424, a mapping from code view register numbers to dwarf, to replace it.

**Comment by labath** [comment]
I don't think we need two mappings. You can use RegisterContext::ConvertBetweenRegisterKinds (or some other function in that area) to convert between the kinds that lldb knows about, so the conversion to lldb numbers should suffice.

That said, the main purpose of that code seems to be the construction of dwarf expressions, so it may also be possible to *replace* it with something which maps onto dwarf register numbers directly (maybe even by making use of some of the tables inside llvm).

**Comment by ZequanWu** [comment]
In my understanding, RegisterContext is not available until the debugged process starts running, and the construction of dwarf location expression happens at parsing debug info phase in symbol file plugin. Also the mapping from dwarf register number to real register is abi-dependent. I'm not sure how can we map into dwarf register numbers correctly.

**Comment by labath** [comment]
You are correct. The symbol file plugin does not have access to a register context, so it cannot use it to convert between register types.

So, I guess we would need to do the mapping ourselves. Ideally we wouldn't have two mappings (the usage of the lldb register scheme is not really correct here, as the process plugin is not required to use those constants for its registers), but rather replace all the uses of the lldb scheme with dwarf registers.

Or we could take this as another sign that we really should stop conjuring up dwarf expressions.

**Comment by ZequanWu** [comment]
> So, I guess we would need to do the mapping ourselves.
Do you mean manually mapping from codeview register to dwarf registers?

**Comment by labath** [comment]
Yes

**Comment by ZequanWu** [comment]
Adding a note for myself.

If we have an abstract layer to describe variable locations, then we can correctly represent the variable locations for the following example in code view. In the range  [0001:1556496, +45), the value of `this` is located at rcx and in the range [0001:1556541, +148), the value of `this` is located at rdi.

Right now, only the first `S_DEFRANGE_REGISTER` is used, causing some variables not accessible even though they should.
```
    2320 | S_LOCAL [size = 16] `this`
           type=0xC0C5 (base::PartitionRoot<1>*), flags = param
    2336 | S_DEFRANGE_REGISTER [size = 16]
           register = RCX, may have no name = false, range start = 0001:1556496, length = 45
           gaps = []
    2352 | S_DEFRANGE_REGISTER [size = 16]
           register = RDI, may have no name = false, range start = 0001:1556541, length = 148
           gaps = []
```
